freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RepairInstancesService.java

```diff
@@ -255,4 +273,37 @@ public OperationStatus rebootInstances(String accountId, RebootInstancesRequest
         }
         return operationToOperationStatusConverter.convert(operation);
     }
+
+    public DescribeFreeIpaResponse rebuild(String accountId, RebuildRequest rebuildRequest) {
+        Stack stack = stackService.findByCrnAndAccountIdWithListsEvenIfTerminated(rebuildRequest.getEnvironmentCrn(), accountId, rebuildRequest.getSourceCrn())
+                .orElse(null);
+        if (stack == null) {
+            String error = String.format("Stack not found with CRN %s in environment %s.", rebuildRequest.getSourceCrn(), rebuildRequest.getEnvironmentCrn());
+            LOGGER.error(error);
+            throw new BadRequestException(error);
+        }
```

Comment:
I would suggest to move the logic to `StackService` like we have it with `getByEnvironmentCrnAndAccountIdEvenIfTerminated`, you would only need the extra joins there, so it could be reusable later

Reply:
done